include/rocksdb/cloud/cloud_file_system.h

```diff
@@ -663,6 +587,117 @@ class CloudFileSystem : public FileSystem {
       const std::string& dbname,
       const std::vector<std::string>& active_cookies) = 0;
 
+  // returns the options used to create this object
+  virtual const CloudFileSystemOptions& GetCloudFileSystemOptions() const = 0;
+
+  // The SrcBucketName identifies the cloud storage bucket and
+  // GetSrcObjectPath specifies the path inside that bucket
+  // where data files reside. The specified bucket is used in
+  // a readonly mode by the associated DBCloud instance.
+  virtual const std::string& GetSrcBucketName() const = 0;
+  virtual const std::string& GetSrcObjectPath() const = 0;
+  virtual bool HasSrcBucket() const = 0;
+
+  // The DestBucketName identifies the cloud storage bucket and
+  // GetDestObjectPath specifies the path inside that bucket
+  // where data files reside. The associated DBCloud instance
+  // writes newly created files to this bucket.
+  virtual const std::string& GetDestBucketName() const = 0;
+  virtual const std::string& GetDestObjectPath() const = 0;
+  virtual bool HasDestBucket() const = 0;
+
+  virtual bool SrcMatchesDest() const = 0;
+
+  // Get the storage provider for the FileSystem.
+  virtual const std::shared_ptr<CloudStorageProvider>& GetStorageProvider()
+      const = 0;
+
+  virtual Logger* GetLogger() const = 0;
+};
+
+//
+// The Cloud file system
+//
+// NOTE: The AWS SDK must be initialized before the CloudFileSystem is
+// constructed, and remain active (Aws::ShutdownAPI() not called) as long as any
+// CloudFileSystem objects exist.
+class CloudFileSystemEnv : public CloudFileSystem {
+ protected:
```

Comment:
curious why do we keep this separate class instead of moving all these static functions and member variables to `CloudFileSystemImpl` and simply removing this `CloudFileSystemEnv` class?

Reply:
I am not sure. I kind of assumed there was a reason for these methods to be in CloudFileSystem in the first place. Maybe somebody didn't want to reveal the fact that ColoudFileSystemImpl exists to the client code or didn't want to force them to add another include.
There is some logic to the current state. CloudFileSystem defines the API, CloudFileSystemEnv has various static initialization methods. Theoretically it could even be made separate from the CloudFileSystem class hierarchy altogether. Finally, CloudFileSystemImp implements the CloudFileSystem and FileSystem APIs.

That being said, I am not opposed to moving everything to CloudFileSystemImpl if that is the consensus.

Reply:
> Theoretically it could even be made separate from the CloudFileSystem class hierarchy altogether. Finally, CloudFileSystemImp implements the CloudFileSystem and FileSystem APIs

Like this idea.